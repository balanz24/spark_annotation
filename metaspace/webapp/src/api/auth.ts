const fetchPostJson = async(url: string, body: object) =>
  fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    credentials: 'include',
    body: JSON.stringify(body),
  })

export const signOut = async() => {
  const response = await fetch('/api_auth/signout', { method: 'POST', credentials: 'include' })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}

export const signInByEmail = async(email: string, password: string): Promise<boolean> => {
  const response = await fetchPostJson('/api_auth/signin', { email, password })
  return response.status >= 200 && response.status < 300
}

export const createAccountByEmail = async(email: string, password: string, name: string) => {
  const response = await fetchPostJson('/api_auth/createaccount', { email, password, name })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}

export const sendPasswordResetToken = async(email: string) => {
  const response = await fetchPostJson('/api_auth/sendpasswordresettoken', { email })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}

export const validatePasswordResetToken = async(token: string, email: string): Promise<boolean> => {
  const response = await fetchPostJson('/api_auth/validatepasswordresettoken', { token, email })
  return response.status === 200
}

export const resetPassword = async(token: string, email: string, password: string) => {
  const response = await fetchPostJson('/api_auth/resetpassword', { token, email, password })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}
