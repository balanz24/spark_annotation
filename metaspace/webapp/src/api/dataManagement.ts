import gql from 'graphql-tag'

export interface GroupListItem {
  id: string;
  name: string;
}

export const oneGroupQuery =
  gql`query oneGroupQuery($groupId: ID!) {
  group(groupId:$groupId) {
    id
    name
  }
}`

export const oneProjectQuery =
  gql`query oneProjectQuery($projectId: ID!) {
  project(projectId:$projectId) {
    id
    name
  }
}`

export const allGroupsQuery =
gql`query allGroupsQuery($query: String) {
  allGroups(query:$query) {
    id
    name
  }
}`

export const requestAccessToGroupMutation =
gql`mutation($groupId: ID!) {
  requestAccessToGroup(groupId: $groupId){
    role
    }
  }`
