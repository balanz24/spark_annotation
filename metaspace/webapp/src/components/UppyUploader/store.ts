// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from '@vue/composition-api'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach(listener => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
