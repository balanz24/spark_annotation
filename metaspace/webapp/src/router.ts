/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/imzml/DatasetBrowserPage'),
  DatasetEnrichmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/enrichment/DatasetEnrichmentPage'),
  SpottingProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ './modules/SpottingProject/DashboardPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ './design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ './design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ './design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ './design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ './design/FormsPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    {
      path: '/datasets/:dataset_id/comparison',
      name: 'datasets-comparison',
      component: asyncPages.DatasetComparisonPage,
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
    { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
    { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
    { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/groups', component: asyncPages.GroupsListPage },
    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
    { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
    { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
    { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
    { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

    { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
