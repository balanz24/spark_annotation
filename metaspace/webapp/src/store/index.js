import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import getters from './getters.js';
import mutations from './mutations.js';
import actions from './actions.js';
import {accountModule} from '../modules/Account';

const store = new Vuex.Store({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: {visible: false},

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE'
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule
  }
})

export default store;
