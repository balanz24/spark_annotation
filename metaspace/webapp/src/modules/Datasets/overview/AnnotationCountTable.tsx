import { defineComponent } from '@vue/composition-api'
import Vue from 'vue'
import { Table, TableColumn, Button } from '../../../lib/element-ui'
import { DatasetAnnotationCount } from '../../../api/dataset'
import { encodeParams } from '../../Filters'

const RouterLink = Vue.component('router-link')

interface AnnotationCountTableData{
  name: string
  [key: string]: any // dynamic keys according to fdrs
}

interface AnnotationCountTableProps {
  id: string
  sumRowLabel: string
  btnLabel: string
  title: string
  data: DatasetAnnotationCount[]
  header: string[]
  headerTitleSuffix: string
}

export const AnnotationCountTable = defineComponent<AnnotationCountTableProps>({
  name: 'AnnotationCountTable',
  props: {
    id: { type: String, default: '' },
    title: { type: String, default: '# of annotations with FDR% <=' },
    sumRowLabel: { type: String, default: 'Total Annotations' },
    btnLabel: { type: String, default: 'Browse annotations' },
    data: { type: Array, default: () => [] },
    header: { type: Array, default: () => [5, 10, 20, 50] },
    headerTitleSuffix: { type: String, default: '%' },
  },
  setup(props, ctx) {
    const annotationsLink = (datasetId: string, database?: string, fdrLevel?: number) => {
      const query = {
        database,
        fdrLevel,
        datasetIds: [datasetId],
      }
      // delete undefined so that filter do not replace the nulls and make the navigation back weird
      if (!database) {
        delete query.database
      }
      if (!fdrLevel) {
        delete query.fdrLevel
      }

      return {
        name: 'dataset-annotations',
        params: { dataset_id: datasetId },
        query: encodeParams(query),
      }
    }

    const mountTableData = (data: DatasetAnnotationCount[], header: string[]) : AnnotationCountTableData[] => {
      if (!data) { return [] }

      return data.map((item: DatasetAnnotationCount) => {
        const obj = {} as AnnotationCountTableData
        obj.id = item?.databaseId
        obj.name = `${item?.dbName}${item?.dbVersion ? `-${item?.dbVersion}` : ''}`
        header.forEach((headerEl: string) => {
          obj[headerEl] = Array.isArray(item?.counts)
            ? (item.counts.find((count: any) => count.level === headerEl) || {}).n : '-'
        })
        return obj
      })
    }

    const browseAnnotations = () => {
      const { id } = props
      ctx.root.$router.push(annotationsLink(id))
    }

    const formatTitle = (col: number|string) => {
      const { headerTitleSuffix } = props
      return `${col === 'name' ? 'Database' : col}${col !== 'name' ? headerTitleSuffix : ''}`
    }

    const formatCell = (row: any, column: any, cellValue: any, rowIndex: number, colIndex: number) => {
      const { id } = props
      if (colIndex === 0) {
        return (
          <RouterLink to={annotationsLink(id?.toString(), row?.id,
            undefined)}>
            {cellValue}
          </RouterLink>
        )
      }

      return (
        <RouterLink to={annotationsLink(id?.toString(), row?.id,
          parseInt(column?.property, 10) / 100)}>
          {cellValue}
        </RouterLink>
      )
    }

    const getSummaries = (param: any) => {
      const { columns, data } = param
      return columns
        .map((col: any, colIndex : number) => {
          if (colIndex === 0) {
            return props.sumRowLabel
          }
          const currentFDR = props.header[colIndex - 1]

          const reducer = (accumulator: number, currentValue: any) => {
            return accumulator + currentValue[currentFDR]
          }

          return (
            <RouterLink key={colIndex} to={annotationsLink(props.id?.toString(), undefined,
              parseInt(currentFDR, 10) / 100)}>
              {data.reduce(reducer, 0)}
            </RouterLink>
          )
        })
    }

    return () => {
      const { data, header, sumRowLabel, btnLabel, title } = props
      const tableData = mountTableData(data, header)

      return (
        <div class="relative text-center">
          <h4>{title}</h4>
          <Table
            data={tableData}
            show-summary={true}
            sum-text={sumRowLabel}
            summary-method={getSummaries}
            style="width: 100%; margin-top: 20px">
            {
              ['name'].concat(header).map((col, colIndex) => {
                return <TableColumn
                  sortable={true}
                  key={colIndex}
                  prop={col.toString()}
                  formatter={(row: any, column: any, cellValue: any, index: number) =>
                    formatCell(row, column, cellValue, index, colIndex)}
                  label={formatTitle(col)} />
              })
            }
          </Table>
          <div class="text-right mt-2">
            <Button onClick={browseAnnotations}>{btnLabel}</Button>
          </div>
        </div>
      )
    }
  },
})
