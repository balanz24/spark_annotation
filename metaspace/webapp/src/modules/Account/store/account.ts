import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
