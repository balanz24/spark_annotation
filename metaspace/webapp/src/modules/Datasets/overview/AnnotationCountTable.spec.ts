import { mount } from '@vue/test-utils'
import router from '../../../router'
import store from '../../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import { AnnotationCountTable } from './AnnotationCountTable'
import { DatasetAnnotationCount } from '../../../api/dataset'

Vue.use(Vuex)
sync(store, router)

describe('AnnotationCountTable', () => {
  const mockAnnotationCounts = [
    {
      databaseId: 1,
      dbName: 'HMDB',
      dbVersion: 'v4',
      counts: [
        {
          level: 5,
          n: 55,
        },
        {
          level: 10,
          n: 104,
        },
        {
          level: 20,
          n: 196,
        },
        {
          level: 50,
          n: 224,
        },
      ],
    },
    {
      databaseId: 6,
      dbName: 'ChEBI',
      dbVersion: '2018-01',
      counts: [
        {
          level: 5,
          n: 65,
        },
        {
          level: 10,
          n: 90,
        },
        {
          level: 20,
          n: 177,
        },
        {
          level: 50,
          n: 190,
        },
      ],
    },
  ]
  const id = 'xxxx'
  const mockFdrLevels = [5, 10, 20, 50]
  const propsData = { data: mockAnnotationCounts, id: id, header: mockFdrLevels }
  const propsEmptyData = { data: [], id: id, header: mockFdrLevels }

  const testHarness = Vue.extend({
    components: {
      AnnotationCountTable,
    },
    render(h) {
      return h(AnnotationCountTable, { props: this.$attrs })
    },
  })

  it('it should match snapshot', async() => {
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('it should render the correct number of columns', async() => {
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()
    expect(wrapper.findAll('th').length).toBe(mockFdrLevels.length + 1)
  })

  it('it should render the correct links to navigate to annotation', async() => {
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    // check if each link cell has the correct parameters for fdr and db filters
    wrapper.find('tbody').findAll('tr').wrappers.forEach((row, rowIndex) => {
      row.findAll('a').wrappers.forEach((col, colIndex) => {
        if (colIndex !== 0) { // db name links do not filter by fdr
          expect(col.props('to').query.fdr)
            .toBe((mockFdrLevels[colIndex - 1] / 100).toString())
        }
        expect(col.props('to').params.dataset_id).toBe('xxxx')
        expect(col.props('to').query.db_id).toBe(mockAnnotationCounts[rowIndex].databaseId.toString())
      })
    })

    // check if each link cell has the correct parameters for the summaryRow
    wrapper.findAll('tbody').at(1)
      .findAll('a').wrappers.forEach((col, colIndex) => {
        expect(col.props('to').query.fdr)
          .toBe((mockFdrLevels[colIndex] / 100).toString())
        expect(col.props('to').params.dataset_id).toBe('xxxx')
      })

    expect(wrapper.findAll('th').length).toBe(mockFdrLevels.length + 1)
    expect(wrapper.findAll('a').at(1).props('to').params.dataset_id)
      .toBe('xxxx')
  })

  it('it should render with no data warning when props empty', async() => {
    const wrapper = mount(testHarness, { store, router, propsData: propsEmptyData })
    await Vue.nextTick()

    expect(wrapper.find('.el-table__empty-text').exists()).toBeTruthy()
  })

  it('it should render the annotation count sum correctly', async() => {
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    // check if each link cell has the correct parameters
    wrapper.find('.el-table__footer-wrapper').findAll('td').wrappers.slice(1)
      .forEach((footerItem, index) => {
        const reducer = (accumulator: number, currentValue: DatasetAnnotationCount) => {
          return accumulator + currentValue.counts[index].n
        }
        const sum = mockAnnotationCounts.reduce(reducer, 0)
        expect(parseInt(footerItem.text(), 10)).toBe(sum)
      })
  })

  it('it should render the correct databases name', async() => {
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    // check if each link cell has the correct parameters
    wrapper.find('.el-table__footer-wrapper').findAll('td').wrappers.slice(1)
      .forEach((footerItem, index) => {
        const reducer = (accumulator: number, currentValue: DatasetAnnotationCount) => {
          return accumulator + currentValue.counts[index].n
        }
        const sum = mockAnnotationCounts.reduce(reducer, 0)
        expect(parseInt(footerItem.text(), 10)).toBe(sum)
      })
  })

  it('it should navigate to dataset annotation page on browse annotation button click', async() => {
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    wrapper.find('button').trigger('click')
    await Vue.nextTick()
    expect(wrapper.vm.$route.path).toBe(`/dataset/${id}/annotations`)
  })
})
