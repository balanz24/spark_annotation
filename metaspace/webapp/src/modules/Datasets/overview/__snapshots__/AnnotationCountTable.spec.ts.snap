// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotationCountTable it should match snapshot 1`] = `
<div class="relative text-center" data="[object Object],[object Object]" id="xxxx" header="5,10,20,50">
  <h4># of annotations with FDR% &lt;=</h4>
  <div class="el-table el-table--fit el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition" style="width: 100%; margin-top: 20px;">
    <div class="hidden-columns">
      <div></div>
      <div></div>
      <div></div>
      <div></div>
      <div></div>
    </div>
    <div class="el-table__header-wrapper">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 400px;">
        <colgroup>
          <col name="el-table_1_column_1" width="80">
          <col name="el-table_1_column_2" width="80">
          <col name="el-table_1_column_3" width="80">
          <col name="el-table_1_column_4" width="80">
          <col name="el-table_1_column_5" width="80">
        </colgroup>
        <thead class="">
          <tr class="">
            <th colspan="1" rowspan="1" class="el-table_1_column_1     is-leaf is-sortable">
              <div class="cell">Database<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_2     is-leaf is-sortable">
              <div class="cell">5%<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_3     is-leaf is-sortable">
              <div class="cell">10%<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_4     is-leaf is-sortable">
              <div class="cell">20%<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_5     is-leaf is-sortable">
              <div class="cell">50%<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div class="el-table__body-wrapper is-scrolling-left">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 400px;">
        <colgroup>
          <col name="el-table_1_column_1" width="80">
          <col name="el-table_1_column_2" width="80">
          <col name="el-table_1_column_3" width="80">
          <col name="el-table_1_column_4" width="80">
          <col name="el-table_1_column_5" width="80">
        </colgroup>
        <tbody>
          <tr class="el-table__row">
            <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=1&amp;ds=xxxx" class="">HMDB-v4</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=1&amp;ds=xxxx&amp;fdr=0.05" class="">55</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=1&amp;ds=xxxx&amp;fdr=0.1" class="">104</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=1&amp;ds=xxxx&amp;fdr=0.2" class="">196</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_5  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=1&amp;ds=xxxx&amp;fdr=0.5" class="">224</a></div>
            </td>
          </tr>
          <tr class="el-table__row">
            <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=6&amp;ds=xxxx" class="">ChEBI-2018-01</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=6&amp;ds=xxxx&amp;fdr=0.05" class="">65</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=6&amp;ds=xxxx&amp;fdr=0.1" class="">90</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=6&amp;ds=xxxx&amp;fdr=0.2" class="">177</a></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_5  ">
              <div class="cell"><a href="/dataset/xxxx/annotations?db_id=6&amp;ds=xxxx&amp;fdr=0.5" class="">190</a></div>
            </td>
          </tr>
          <!---->
        </tbody>
      </table>
      <!---->
      <!---->
    </div>
    <div class="el-table__footer-wrapper">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__footer" style="width: 400px;">
        <colgroup>
          <col name="el-table_1_column_1" width="80">
          <col name="el-table_1_column_2" width="80">
          <col name="el-table_1_column_3" width="80">
          <col name="el-table_1_column_4" width="80">
          <col name="el-table_1_column_5" width="80">
        </colgroup>
        <tbody class="">
          <tr>
            <td colspan="1" rowspan="1" class="el-table_1_column_1 is-leaf">
              <div class="cell">Total Annotations</div>
            </td>
            <td colspan="1" rowspan="1" class="el-table_1_column_2 is-leaf">
              <div class="cell"><a href="/dataset/xxxx/annotations?ds=xxxx&amp;fdr=0.05" class="">120</a></div>
            </td>
            <td colspan="1" rowspan="1" class="el-table_1_column_3 is-leaf">
              <div class="cell"><a href="/dataset/xxxx/annotations?ds=xxxx&amp;fdr=0.1" class="">194</a></div>
            </td>
            <td colspan="1" rowspan="1" class="el-table_1_column_4 is-leaf">
              <div class="cell"><a href="/dataset/xxxx/annotations?ds=xxxx&amp;fdr=0.2" class="">373</a></div>
            </td>
            <td colspan="1" rowspan="1" class="el-table_1_column_5 is-leaf">
              <div class="cell"><a href="/dataset/xxxx/annotations?ds=xxxx&amp;fdr=0.5" class="">414</a></div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <!---->
    <!---->
    <div class="el-table__column-resize-proxy" style="display: none;"></div>
  </div>
  <div class="text-right mt-2"><button type="button" class="el-button el-button--default">
      <!---->
      <!----><span>Browse annotations</span></button></div>
</div>
`;
